### DISCUSSION TITLE
Await storage setItem in persist middleware

### DISCUSSION THREAD

**Discussion by sumartoyo** [pmndrs/zustand#287] [Q&A] [answered]
setItem call in persist middleware is not await-ed. But some storage may do the setItem asynchronously.

[Code](https://github.com/pmndrs/zustand/blob/0ae00d7760935be0ea894dc3824859d2de9fbcc5/src/middleware.ts#L219) from middleware.ts

```js
  const setItem = async () => {
    const state = { ...get() }

    if (whitelist) {
      Object.keys(state).forEach((key) => {
        !whitelist.includes(key) && delete state[key]
      })
    }
    if (blacklist) {
      blacklist.forEach((key) => delete state[key])
    }

    storage?.setItem(name, await serialize({ state, version }))
  }
```

Is it intentional? If it is then how to catch error in case an exception is thrown when calling setItem?

**Comment by AnatoleLucet** [comment, marked as answer]
I don't see why we'd need to await `setItem` nor catch an error related to it. Can you give an example?

**Reply by sumartoyo** [reply]
One example is when we use AsyncStorage for React Native. https://react-native-async-storage.github.io/async-storage/docs/api/#setitem. There is a possibility of underlying storage error. A common case is when the storage size has reached its limit.

As for the await itself, in my experience maybe only one or two cases that I need to await it. For example, when logging out a user, I would await it to make sure the persisted login-related state is cleared. Generally, it is to make sure that the state is persisted.

**Reply by AnatoleLucet** [reply]
> There is a possibility of underlying storage error.

Sure, but still, why should this middleware catch this error?

> For example, when logging out a user, I would await it to make sure the persisted login-related state is cleared. Generally, it is to make sure that the state is persisted.

So something similar to [`onRehydrateStorage`](https://github.com/pmndrs/zustand/blob/0ae00d7760935be0ea894dc3824859d2de9fbcc5/src/middleware.ts#L158-L163) (but for setters) might be what you need for this?

**Reply by sumartoyo** [reply]
Oh sorry my bad. I didn't mean this middleware should catch the error. What I meant is the error should be bubbled up to the caller.

This is currently what will happen.
```js
export const useStore = create(persist(
  (set, get) => ({
    fishes: 0,
    addAFish: async () => {
      try {
        await set({ fish: get().fish + 1 })
      } catch (err) {
        // error of storage.setItem() will never get here because it runs asynchronously, not awaited
      }
    }
  }),
  {
    name: "food-storage",
    getStorage: () => AsyncStorage,
  }
))
```

>So something similar to onRehydrateStorage (but for setters) might be what you need for this?

Yes it will suit my need. Another way we can do it is by replacing [middleware.ts#L219](https://github.com/pmndrs/zustand/blob/0ae00d7760935be0ea894dc3824859d2de9fbcc5/src/middleware.ts#L219) with  
`return storage?.setItem(name, await serialize({ state, version }))`  
and [middleware.ts#L226](https://github.com/pmndrs/zustand/blob/0ae00d7760935be0ea894dc3824859d2de9fbcc5/src/middleware.ts#L226) with  
`return setItem()`  
In that way, the error from `storage.setItem()` will be "bubbled up" (I don't know a better term for it) to the caller.

But before going to code changes, I want to ask first if is it intentionally implemented that way? I mean maybe there are reasons why the error is not bubbled up to the caller that I miss here?

**Reply by AnatoleLucet** [reply]
Indeed, I didn't think about that.

> Another way we can do it is by replacing [middleware.ts#L219](https://github.com/pmndrs/zustand/blob/0ae00d7760935be0ea894dc3824859d2de9fbcc5/src/middleware.ts#L219) with
> `return storage?.setItem(name, await serialize({ state, version }))`
> and [middleware.ts#L226](https://github.com/pmndrs/zustand/blob/0ae00d7760935be0ea894dc3824859d2de9fbcc5/src/middleware.ts#L226) with
> `return setItem()`

You're right, it'd be easier to use.

> But before going to code changes, I want to ask first if is it intentionally implemented that way?

No, it's not really intentional. I just didn't think of this use case.

**Reply by sumartoyo** [reply]
I see. So we can fix it if it is not intentional.

**Reply by AnatoleLucet** [reply]
Sure, feel free to open a PR if you want :slightly_smiling_face:
